### PR TITLE
Enable `x.py check` for miri

### PR DESCRIPTION
**Pull request by oli-obk**
Now that the miri subtree is working properly, let's add it to x.py check.

cc @rust-lang/miri

### CODE REVIEW

File: src/bootstrap/tool.rs [hunk 1]
```diff
@@ -865,17 +860,17 @@ macro_rules! tool_extended {
...
-    Cargofmt, "src/tools/rustfmt", "cargo-fmt", stable=true, in_tree=true, {};
-    CargoClippy, "src/tools/clippy", "cargo-clippy", stable=true, in_tree=true, {};
-    Clippy, "src/tools/clippy", "clippy-driver", stable=true, in_tree=true, {};
-    Miri, "src/tools/miri", "miri", stable=false, in_tree=true, {};
-    CargoMiri, "src/tools/miri/cargo-miri", "cargo-miri", stable=false, in_tree=true, {};
+    Cargofmt, "src/tools/rustfmt", "cargo-fmt", stable=true, {};
+    CargoClippy, "src/tools/clippy", "cargo-clippy", stable=true, {};
+    Clippy, "src/tools/clippy", "clippy-driver", stable=true, {};
+    Miri, "src/tools/miri", "miri", stable=true, {};
+    CargoMiri, "src/tools/miri/cargo-miri", "cargo-miri", stable=true, {};
```
Comment (Urgau) [inline src/bootstrap/tool.rs:867]: This switch miri from unstable to stable, probably not what you wanted.
```suggestion
    Miri, "src/tools/miri", "miri", stable=false, {};
    CargoMiri, "src/tools/miri/cargo-miri", "cargo-miri", stable=false, {};
```

**Comment by RalfJung** [comment]
r? @RalfJung 
r=me when CI is green

Was this the thing that caused doc failures? We might have to add that recursion limit attribute the error message was talking about...

File: src/bootstrap/tool.rs [hunk 2]
```diff
@@ -795,7 +795,6 @@ macro_rules! tool_extended {
        $path:expr,
        $tool_name:expr,
        stable = $stable:expr,
-       $(in_tree = $in_tree:expr,)?
        $(tool_std = $tool_std:literal,)?
        $extra_deps:block;)+) => {
```
Comment (jyn514) [inline src/bootstrap/tool.rs:799]: looks like `extra_deps` is never used, can we remove it?

**Comment by matthiaskrgr** [comment]
@bors r- rollup=iffy failed in a rollup
https://github.com/rust-lang/rust/pull/103097

**Comment by RalfJung** [comment]
```
error: unclosed HTML tag `T`
  --> src/tools/miri/src/concurrency/weak_memory.rs:20:16
   |
20 | //! std::atomic<T> API). It is therefore possible for this implementation to generate behaviours never observable when the
   |                ^^^
   |
   = note: `-D rustdoc::invalid-html-tags` implied by `-D warnings`
help: try marking as source code
   |
20 | //! `std::atomic<T>` API). It is therefore possible for this implementation to generate behaviours never observable when the
   |     +              +

error: could not document `miri`
```
Needs an upstream Miri sync.

**Comment by RalfJung** [comment]
Looking at the build logs, this does not seem to check cargo-miri. Is that deliberate?

**Comment by matthiaskrgr** [comment]
I don't think we do x.py check in ci, just x.py build/test/dist? 🤔

**Comment by RalfJung** [comment]
"mingw-check" in PR CI does a check build, I think. That's where I checked the logs. It would be awfully misnamed otherwise. ;)

**Comment by RalfJung** [comment]
Miri sync landed, so this should work now.
@bors r+

**Comment by RalfJung** [comment]
Interesting, usually warnings during the Miri sysroot build are fine but here something is adding a `-D warnings` it seems. Looks like bootstrap injects these flags *after* cargo, rather than the usual way of setting RUSTFLAGS before invoking cargo?

**Comment by Urgau** [comment]
@RalfJung Yes for linting bootstrap use flags not environment variables, see this (note the `arg` instead of `env`): https://github.com/rust-lang/rust/blob/a5406feb1cf1a9f2479b72917cf4225e5f6aa240/src/bootstrap/builder.rs#L1888-L1900

**Comment by RalfJung** [comment]
Yeah and that is then injected to rustc here

https://github.com/rust-lang/rust/blob/9110d925d0b3e4842376e830f4404a28e1aa2658/src/bootstrap/bin/rustc.rs#L80-L82

bypassing cargo/RUSTFLAGS.

File: src/bootstrap/check.rs [hunk 3]
```diff
@@ -451,14 +451,12 @@ macro_rules! tool_check_step {
...
-// Clippy and Rustfmt are hybrids. They are external tools, but use a git subtree instead
+// Clippy, miri and Rustfmt are hybrids. They are external tools, but use a git subtree instead
 // of a submodule. Since the SourceType only drives the deny-warnings
 // behavior, treat it as in-tree so that any new warnings in clippy will be
 // rejected.
 tool_check_step!(Clippy, "src/tools/clippy", SourceType::InTree);
-// Miri on the other hand is treated as out of tree, since InTree also causes it to
-// be run as part of `check`, which can fail on platforms which libffi-sys has no support for.
-tool_check_step!(Miri, "src/tools/miri", SourceType::Submodule);
+tool_check_step!(Miri, "src/tools/miri", SourceType::InTree);
```
Comment (RalfJung) [inline src/bootstrap/check.rs:459]: @oli-obk can you add cargo-miri here?

**Comment by RalfJung** [comment]
All right, time for another attempt.
@bors r+

**Comment by RalfJung** [comment]
Ah, that error is back
```
  = help: consider increasing the recursion limit by adding a `#![recursion_limit = "256"]` attribute to your crate (`miri`)
```
No idea why this occurs here, when actually generating the documentation for https://doc.rust-lang.org/nightly/nightly-rustc/miri/index.html works perfectly fine. But maybe we should just add that attribute?

**Comment by RalfJung** [comment]
Okay, let's do that then -- I pushed it to this PR.
@bors r+

**Pull request merged**